marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/entity-lib.sjs

```diff
@@ -0,0 +1,140 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+'use strict';
+
+/**
+ * This library is intended to encapsulate all logic specific to Entity Services models. As of DHF 5.2.0, this logic is
+ * spread around the DH codebase. It is expected that this will gradually be refactored so that all ES-specific logic
+ * resides in this module to promote reuse and also simplify upgrades as Entity Services changes within MarkLogic.
+ */
+
+const sem = require("/MarkLogic/semantics.xqy");
+const semPrefixes = {es: 'http://marklogic.com/entity-services#'};
+
+/**
+ * @return an array of strings, one for each EntityType
+ */
+function findEntityTypeIds() {
+  return cts.triples(null,
+    sem.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
+    sem.iri('http://marklogic.com/entity-services#EntityType')
+  ).toArray().map(triple => sem.tripleSubject(triple).toString());
+}
+
+/**
+ * @return a map object where each key is an EntityTypeId, and the value of each key is the corresponding entity type
+ */
+function findEntityTypesAsMap() {
+  const map = {};
+  for (var doc of cts.search(cts.collectionQuery("http://marklogic.com/entity-services/models"))) {
+    Object.assign(map, convertModelToEntityTypeMap(doc.toObject()));
+  }
+  return map;
+}
+
+/**
+ * @param entityTypeId string or a sem.iri
+ * @return the matching model document, or null if one is not found
+ */
+function findModelForEntityTypeId(entityTypeId) {
+  return fn.head(cts.search(
+    cts.andQuery([
+      cts.collectionQuery('http://marklogic.com/entity-services/models'),
+      cts.tripleRangeQuery(sem.iri(entityTypeId), sem.curieExpand("rdf:type"), sem.curieExpand("es:EntityType", semPrefixes))
+    ])));
+}
+
+/**
+ * @param entityTypeId string or a sem.iri
+ * @return null if a model can't be found matching the given EntityTypeId, or if a model is found but there's no an entity type
+ * with a title matching the entity title in the EntityTypeId. Otherwise, the entity type from the definitions array in the model
+ * is returned.
+ */
+function findEntityType(entityTypeId) {
+  const modelDoc = findModelForEntityTypeId(entityTypeId);
+  if (!modelDoc) {
+    return null;
+  }
+  return modelDoc.toObject().definitions[getEntityTypeIdParts(entityTypeId).entityTypeTitle];
+}
+
+/**
+ * @param entityTypeId sem.iri or string
+ * @return {{entityTypeTitle: string, baseUri: string, modelTitle: string, version: string}}
+ */
+function getEntityTypeIdParts(entityTypeId) {
+  if (entityTypeId == null) {
+    throw Error("Cannot get EntityTypeId parts from null EntityTypeId");
+  }
+
+  entityTypeId = entityTypeId.toString();
+
+  const tokens = entityTypeId.split("/");
+  if (tokens.length < 3) {
+    throw Error("Could not get EntityTypeId parts from invalid EntityTypeId: " + entityTypeId);
+  }
+
+  const infoTokens = tokens[tokens.length - 2].split("-");
```

Comment:
This still has the issue of not working if the EntityTypeTitle has a slash in it - which seems very unlikely. There's a test to verify that it fails. 

@ryanjdew you had mentioned grabbing the model doc and figuring the baseUri from that. I'd like for this method to not avoid any DB calls though - in fact, I have a convention of "find" implying the need to query ML, and "get" not involving any queries (I've used that convention in the past, it's a handy way for a client to know how performant a function should be).